script/windows/14.0/update/WindowsUpdateCheck.php

```diff
@@ -0,0 +1,161 @@
+      if(!empty($desktop_update['bundle'])) {
+        $newAppVersion = $desktop_update['bundle']->version;
+      } else {
+        $newAppVersion = $appVersion;
+      }
+      $desktop_update['keyboards'] = $this->BuildKeyboardsResponse($newAppVersion, $packages);
+
+      return json_encode($desktop_update, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
+    }
+
+    private function BuildKeymanDesktopVersionResponse($tier, $InstalledVersion, $regex) {
+      if(empty($this->DownloadVersions)) {
+        $this->DownloadVersions = DownloadsApi::Instance()->GetPlatformVersion("windows");
+        if($this->DownloadVersions === NULL) {
+          fail('Unable to download or decode version data from '.KeymanHosts::Instance()->downloads_keyman_com, 500);
+        }
+
+        if(!isset($this->DownloadVersions->windows)) {
+          fail("Unable to find {windows} key in ".KeymanHosts::Instance()->downloads_keyman_com." data", 500);
+        }
+      }
+
+      // Check each of the tiers for the one that matches the major version.
+      // This gets us upgrades on alpha, beta and stable tiers.
+      $tiers = get_object_vars($this->DownloadVersions->windows);
+
+      return $this->CheckVersionResponse($tier, $tiers, $InstalledVersion, $regex);
+    }
+
+    private function CheckVersionResponse($tier, $tiers, $InstalledVersion, $regex) {
+      if(!isset($tiers[$tier])) return FALSE;
+      $tierdata = $tiers[$tier];
+
+      // We will support staying on alpha or beta tier once a version
+      // hits stable. This is correct, with major upgrade support, as
+      // that'll mean the user will transition to a new version on the
+      // same tier (i.e. always staying bleeding edge)
+
+      $files = get_object_vars($tierdata->files);
+      foreach($files as $file => $filedata) {
+        // This is currently tied to Windows -- for other platforms we need to change this
+        if(preg_match($regex, $file)) {
+          $filedata->url = KeymanHosts::Instance()->downloads_keyman_com . "/windows/$tier/{$filedata->version}/{$file}";
+          return $filedata;
+        }
+      }
+
+      return FALSE;
+    }
+
+    private function IsSameMajorVersion($v1, $v2) {
+      if(empty($v1) || empty($v2)) return FALSE;
+      $v1 = explode('.', $v1);
+      $v2 = explode('.', $v2);
+      return $v1[0] == $v2[0];
+    }
+
+    private function BuildKeyboardsResponse($appVersion, $packages) {
+      $keyboards = [];
+
+      // For each keyboard in the parameter request, check for a new version or
+      // for a keyboard that replaces it
+
+      foreach ($packages as $id => $version) {
+        $keyboard = $this->BuildKeyboardResponse($id, $version, $appVersion);
+        if($keyboard !== FALSE) {
+          $keyboards[$id] = $keyboard;
+        }
+      }
+      return $keyboards;
+    }
+
+    private function BuildKeyboardResponse($id, $version, $appVersion) {
+      // TODO: this should use the class instead of file_get_contents
```

Comment:
I assume the `// TODO:` is for later work?

Reply:
Yes. It's not urgent, but an unnecessary http call to the same host seems wasteful.